### ISSUE TITLE
Posting data

### ISSUE THREAD

**Issue by keengene** [tsjost/Binero-php#1]
Total shot in the dark as there haven't been any activity here in years, but is there a possibility of adding POST functionality to this? I would be happy to fork and figure out how to do it myself if someone could point me in the right direction. I'm currently too inexperienced in this field to understand whether the API supports it or not, but wouldn't mind learning something new.

**Comment by tsjost** [comment]
I'm not sure exactly what you mean by adding POST functionality. The code here already supports performing POST requests, as the `Token` API endpoint for authenticating and getting an auth token is done over POST. The second argument of the `call` method in the `Connector` interface is used to specify the HTTP method (GET or POST).

As for adding actual functionality to the user-facing library, like for example being able to purchase domains or pay invoices, that would require reverse engineering Binero's mobile apps to figure out what the API endpoints and their parameters are, as they sadly don't make this information public.

Please provide some more details on what you'd like to accomplish and I'd be happy to help.

**Comment by keengene** [comment]
Sorry! I realized that the original post was a bit vague and confused as to what I actually was trying to achieve and I thought I edited it but apparently that was only in my mind..

The dream is to be able to create, modify and remove DNS records from an automation platform for use in Let's Encrypt certificate validation. Binero is quite lacking in this area at the moment and have only recently made available certificate requests for sites hosted by them.

Or am I missing something very obvious?

As far as reverse engineering, I got to thinking if some packet captures of creating and deleting a few subdomains wouldn't reveal the endpoints the mobile app uses. Or does the app encrypt the information before it transmits? I'm going to try it and find out.

**Comment by tsjost** [comment]
> Or am I missing something very obvious?

Not that I know of, Binero is pretty behind the times.

> ...packet captures of creating and deleting a few subdomains wouldn't reveal the endpoints the mobile app uses. Or does the app encrypt the information before it transmits?

A regular packet capture would just capture encrypted data, yes, so you'll have to perform a [man in the middle attack](https://en.wikipedia.org/wiki/Man-in-the-middle_attack) on yourself -- making the app connect to your fake server which will log the data and forward it to the real server. Packet capture mobile apps should have this built-in.

However if the app has implemented certificate pinning (only allowing the app to communicate with a certificate to which you have no private keys), it becomes a bit more complex, as you'd have to modify the app or root the device to work around the pinning -- or just do static analysis by reading/decompiling the bytecode.

**Comment by fspijkerman** [comment]
Their app has implemented certificate pinning, but its fairly easy to decompile the android app (using apktool for example) 
I share the dream of being able to automate their dns, sadly I haven't found any api calls that allows this.

This dump is based on their Android app version 1.016.
```
smali/com/binero/android/BnrApplication.smali:.field public static final BINERO_URI_DOWNLOAD_INVOICE:Ljava/lang/String; = "https://mobileapi.binero.se/api/Invoices/DownloadInvoicePdf?invoiceNumber=%s"
smali/com/binero/android/rest/RequestParamsBuilder.smali:        
 "api/Account/Logout"
 "api/Account/SubscribeForAndroidPushNotifications?deviceId=%s"
 "api/Account/UserInfo"
 "api/Domains/EditDomainNS?domainName=%s"
 "api/Domains/FindAvailableDomainNames?domainName=%s&campaignCode=%s"
 "api/Domains/GetDomainDetails?domainName=%s"
 "api/Domains/GetDomainList?search=%s&page=%s"
 "api/Domains/GetSupportedTLDs"
 "api/Domains/GetTransferPrices?campaignCode=%s"
 "api/Domains/OrderDomain?campaignCode=%s"
 "api/Domains/RenewDomain?domainName=%s"
 "api/Domains/TransferDomain?campaignCode=%s"
 "api/Domains/TransferDomains?campaignCode=%s"
 "api/Invoices/DownloadInvoicePdf?invoiceNumber=%s"
 "api/Invoices/EditInvoice?invGuid=%s"
 "api/Invoices/GetInvoiceDetails?invGuid=%s"
 "api/Invoices/GetInvoiceList?page=%s&filter=%s"
 "api/News/EditNewsSubscriptions?lang=%s"
 "api/News/GetNewsList?lang=%s&page=%s"
 "api/News/GetNewsPageViewModel?lang=%s"
 "api/Statistics/GetActionsHistory?page=%s&lang=%s"
 "api/Statistics/GetLoginHistory?page=%s"
 "api/Statistics/GetPageViewModel"
 "api/Statistics/GetPageViewModelForPeriod?fromDate=%s&toDate=%s"
 "api/Support/CloseTicket?ticketId=%s"
 "api/Support/CreateReply?ticketId=%s&message=%s"
 "api/Support/CreateTicket?subject=%s&message=%s&department=%s"
 "api/Support/GetTicketDetails?ticketId=%s"
 "api/Support/GetTicketsList?search=%s&page=%s&statusFilter=%s"
```
Hope this might help anyone!

**Comment by tsjost** [comment]
Thanks for the dump!

I actually installed the app and played around a bit, but I can't seem to find any feature to update DNS records? Only modifying name servers. That might explain why there's nothing related to it in the symbols dump :confused: Have you both found an app feature that I haven't?

![image](https://user-images.githubusercontent.com/1737876/54197496-63ca6c80-44c4-11e9-84c4-35620f9ccefe.png)
![image](https://user-images.githubusercontent.com/1737876/54197498-66c55d00-44c4-11e9-84ad-21644735996b.png)

@keengene @fspijkerman maybe we should ask Binero to implement DNS editing in their app? Although it's not been updated since 2014...

Another approach would be to interface against their actual web platform "API", but it's heavily built on Microsoft corporate C# enterprise XML.NET products that I gave up on trying to reverse engineer many years ago; maybe it's worth giving it another shot!
